cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -47,11 +51,144 @@ public CasePurgeFilter(IStorageUtilityIndexed<Case> caseStorage) {
+                    if(existing.getTarget().equals(index.getTarget())) {
+                        if(existing.getRelationship().equals(CaseIndex.RELATIONSHIP_EXTENSION) && !index.getRelationship().equals(CaseIndex.RELATIONSHIP_EXTENSION)) {
+                            toReplace = existing;
+                        } else {
+                            skip = true;
+                        }
+                        break;
+                    }
+                }
+                if(toReplace != null) { 
+                    indexHolder.removeElement(toReplace);
+                }
+                if(!skip) { 
+                    indexHolder.addElement(index);
+                }
+            }
+            String nodeStatus;
+            //Four cases, applied in order. One and two: closed or unowned it starts life dead
+            if(!owned || c.isClosed()) {
+                nodeStatus = STATUS_DEAD;
+            } else {
+                //Otherwise we need to see whether this case maintains any extension indices,
+                //if so, it's abandoned. Otherwise it's Alive
+                boolean abandoned = false;
+                for(CaseIndex index : indexHolder) {
+                    if(index.getRelationship().equals(CaseIndex.RELATIONSHIP_EXTENSION)) {
+                        abandoned = true;
+                        break;
+                    }
+                }
+                if(abandoned) {
+                    nodeStatus = STATUS_ABANDONED;
+                } else {
+                    nodeStatus = STATUS_LIVE;
+                }
+            }
+            
+             
+            g.addNode(c.getCaseId(), new String[] {nodeStatus, String.valueOf(c.getID())});
+            
+            for(CaseIndex index : indexHolder) {
+                g.setEdge(c.getCaseId(), index.getTarget(), index.getRelationship());
+            }
+            indexHolder.removeAllElements();
+        }
+        
+        //Pass 2: Start at Sources (nodes which only produce edges) and walk up
+        //the tree, marking relevant nodes as alive
+        Stack<String> toProcess = g.getSources();
+        while(!toProcess.isEmpty()) {
+            //current node
+            String index = toProcess.pop();
+            String[] node = g.getNode(index);
+            
+            //Walk all indexed nodes, adding them to the process stack.
+            //and update their liveness if necessary.
+            for(Edge<String, String> edge : g.getChildren(index) ) {
+                if(node[0].equals(STATUS_LIVE)) {
+                    g.getNode(edge.i)[0] = STATUS_LIVE;
+                }
+                //add this to the process stack
+                //TODO: We're going to walk stuff redundantly this
+                //way. Can we optimize? 
+                //If parent is already live, does 
+                //that necessarily denote that its ancestors are alive?
+                toProcess.addElement(edge.i);
```

Comment:
Checking to see if edge.i already exists in toProcess seems like a low-hanging optimization.